File: algorithms/js/stringToIntegerAtoi/stringToIntegerAtoi.js
// Source : https://leetcode-cn.com/problems/string-to-integer-atoi/
// Author : Tian Zhuo
// Date   : 2018-11-23

/**
 * @param {string} str
 * @return {number}
 */
var myAtoi = function(str) {
    if(!str || !str.trim()) return 0;
    const num = "1234567890";
    const sign = "+-";
    str = str.trim();
    if((num.indexOf(str[0]) == -1 && sign.indexOf(str[0]) == -1) || (sign.indexOf(str[0]) > -1 && num.indexOf(str[1]) == -1)) return 0;
    for (let i = 1; i < str.length; i++) {
        if(num.indexOf(str[i]) == -1){
            str = str.substring(0, i);
        }
    }
    let res = parseInt(str);
    if(res > 2 ** 31 - 1) return 2 ** 31 - 1;
    if(res < -(2 ** 31)) return -(2 ** 31);
    return res;
};

/**
 * @param {string} str
 * @return {number}
 */
var myAtoi1 = function(str) {
    var test = /^\s*([\+\-]?[0-9]+)/g.exec(str);
    if (!test) return 0;
    
    var res = test[test.length - 1];
    if (res > 2 ** 31 - 1) return 2 ** 31 - 1;
    if (res < -(2 ** 31)) return -(2 ** 31);
    return +res;
};

File: src/main/java/stringToIntegerAtoi/StringToIntegerAtoi.java
/*
 * @Source: https://leetcode-cn.com/problems/string-to-integer-atoi/
 * @Author: Tian Zhuo
 * @Date: 2018-12-13 20:47:26
 */

package stringToIntegerAtoi;

public class StringToIntegerAtoi {
    public int myAtoi(String str) {
        str = str.trim();
        if (str == null || str.length() == 0) {
            return 0;
        }
        final int MAX_DIV = Integer.MAX_VALUE / 10;
        final int MIN_DIV = -(Integer.MIN_VALUE / 10);
        final int MAX_M = Integer.MAX_VALUE % 10;
        final int MIN_M = - (Integer.MIN_VALUE % 10);
        int result = 0;
        int i = 0, len = str.length();
        int sign = 1;
        int digit = str.charAt(0);
        if (digit == '-' || digit == '+') {
            if (digit == '-') {
                sign = -1;
            }
            if (len == 1) {
                return 0;
            }
            i++;
        }
        while (i < len) {
            digit = str.charAt(i++) - '0';
            if (digit >= 0 && digit <= 9) {
                if (sign > 0 && (result > MAX_DIV || (result == MAX_DIV && digit > MAX_M))) { //正数溢出
                    return Integer.MAX_VALUE;
                } else if (sign < 0 && (result > MIN_DIV || (result == MIN_DIV && digit > MIN_M))) { //负数溢出
                    return Integer.MIN_VALUE;
                }
                result = result * 10 + digit;
            } else {
                break;
            }
        }
        return sign > 0 ? result : -result;
    }
}